feat: add XM430 support, table parameterisation, and raw read/write API

jimsynz opened the pull request:
## Summary

This PR adds several enhancements to support a broader range of Robotis servos and use cases:

### XM430 Control Table Support
- Added control table for XM430 series servos (W210, W350) using Protocol 2.0
- Position resolution: 4096 (0-4095 for 0-360 degrees)
- Supports baud rates from 9600 to 4.5 Mbps

### Parameterised Servo Type
- `ControlTable` functions now accept an optional `table_type` parameter
- Allows runtime selection of control table (`:xl330_m288` or `:xm430_w210`)
- Backward compatible - existing arity-1 functions default to `:xl330_m288`
- GenServer accepts `:control_table` option to specify servo type

### Raw Read/Write API
- `read_raw/3` - Read parameter values without unit conversion
- `write_raw/5` - Write raw integer values directly to registers
- Useful for accessing control table registers when automatic unit conversion is not desired

### Error Handling
- Handle empty response case in read operations, returning `{:error, :empty_response}`

All changes are backward compatible with the existing API.

jimsynz committed:
feat: add XM430 control table and parameterise servo type

- Add XM430 series control table (W210, W350) with Protocol 2.0 support
- Update ControlTable module to accept table_type parameter
- Update Robotis GenServer to accept control_table option (defaults to :xl330_m288)
- Add xm430_decode_shutdown and xm430_encode_shutdown functions
- Update dependencies to newer versions (credo 1.7, dialyxir 1.4, etc.)
- Maintain backwards compatibility with existing API (arity-1 functions still work)

XM430 key specs:
- Position resolution: 4096 (0-4095 for 0-360 degrees)
- Baud rates: 9600 to 4.5Mbps
- Current unit: 2.69mA

jimsynz committed:
feat: add read_raw/write_raw functions and improve error handling

- Add `read_raw/3` to read parameter values without unit conversion
- Add `write_raw/5` to write raw integer values directly to registers
- Handle empty response case in read operations, returning `{:error, :empty_response}`

The raw functions are useful for accessing control table registers directly
when automatic unit conversion is not desired.

pkinney commented on lib/robotis.ex:
```diff
@@ -63,9 +85,10 @@ defmodule Robotis do
   def init(opts) do
     port = Keyword.fetch!(opts, :uart_port)
     baud = Keyword.get(opts, :baud, 57_600)
+    control_table = Keyword.get(opts, :control_table, :xl330_m288)
```
I think we don't need to default to anything, and instead we should make the control_table a required option.  What do you think?

pkinney commented on lib/robotis/control_table/tables.ex:
```diff
@@ -136,6 +136,139 @@ defmodule Robotis.ControlTable.Tables do
+  # XM430 series (W210, W350) - Protocol 2.0
+  # Control table is nearly identical to XL330, same addresses and scaling
+  # Main differences: model number, current limit unit (2.69mA vs 1mA)
+  @xm430 %{
```
n.b. We might want to split these into different files just to make it more manageable.

pkinney commented:
@jimsynz Thanks for the addition. I don't have an XM430 to test with, but it's encouraging that getting it to work looks like just changing the control table.

This repo is light on tests and I'm not sure how useful it is to add tests for each new servo, since we would just be testing the control table and comms logic.

pkinney commented:
I'm going to remove the original `encode_param/2 and `decode_param/2` functions.  I think they are only used in a test.

jimsynz committed:
refactor: make `control_table` required and convert to behaviour modules

- Make `control_table` a required option instead of defaulting to `:xl330_m288`
- Add `Robotis.ControlTable` behaviour with `table/0` callback
- Split control tables into separate modules:
  - `Robotis.ControlTable.XL330`
  - `Robotis.ControlTable.XM430`
  - `Robotis.ControlTable.XL320`
- Update tests to provide control table module explicitly

jimsynz commented:
Hey @pkinney thanks for the review. I've updated it by splitting the different control tables into modules and defining a very simple behaviour for them to implement. It was super easy to get it working with other servo models. Thanks!

pkinney commented:
I like that control table module approach. That way someone could implement their own control table without having to add it to the repo.

pkinney commented:
I'll merge it in then fix the formatting issue and update the README.

Now that we have control table modules, we can move those specific functions into them to separate them better.  I'll do that as well.

jimsynz committed:
chore: format control_table.ex
